http-client/src/main/java/io/micronaut/http/client/NettyClientHttpRequest.java

```diff
@@ -180,20 +205,35 @@ protected QueryStringDecoder createDecoder(String uri) {
      */
     HttpRequest getFullRequest(ByteBuf content) {
         String uriStr = resolveUriPath();
-        io.netty.handler.codec.http.HttpMethod method = io.netty.handler.codec.http.HttpMethod.valueOf(httpMethod.name());
+        io.netty.handler.codec.http.HttpMethod method = getMethod(httpMethodName);
         DefaultFullHttpRequest req = content != null ? new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, method, uriStr, content) :
             new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, method, uriStr);
         req.headers().setAll(headers.getNettyHeaders());
         return req;
     }
 
+    /**
+     * This method guarantees that for each name you get at most one instance, so that multiple calls
+     * would not cause GC problems. However, due to the bug in JDK 8, it had to be implemented in a weird way.
+     * @param httpMethodName The name of the method
+     * @return The instance of HttpMethod
+     * @see @link https://bugs.openjdk.java.net/browse/JDK-8161372
+     */
+    private static io.netty.handler.codec.http.HttpMethod getMethod(String httpMethodName) {
+        if (!METHOD_MAP.containsKey(httpMethodName)) {
```

Comment:
This method isn't thread safe

Reply:
You mean ConcurrentHashMap.containsKey is not thread safe? I assumed it's free of locks, unlike computeIfAbsent in case of the key actually present. So, my idea was:

- At startup different threads can block each other for a while when calling computeIfAbsent - but it thread safe, it will not produce inconsistent results.
- After all the possible http methods run at least once, all threads will just call containsKey and get for METHOD_MAP, and these methods are lock free and don't change anything.

Well, I can do this in another way - by modifying DefaultRouteBuilder which actually stores the possible routes and therefore knows all the methods that are available. And he can provide a map to another singleton bean, I guess, which can be used here. All the standard http methods should be in the map regardless of the actual routes, so that would not change the behaviour of the server for them.

Reply:
Let me clarify a little bit more what we are doing here. Basically the most obvious choice is to compute everything directly:

```java
return METHOD_MAP.computeIfAbsent(httpMethodName, io.netty.handler.codec.http.HttpMethod::valueOf);
```

Seems like pretty straightforward. But there is a catch here - in Java 8 this method puts a lock even if the key is present in the map (they fixed that in 9). So multiple requests will create a bottleneck. So I did it like this just so that people using Micronaut with java 8 would not get sudden performance issues. containsKey does not create any locks. If it returns false, we call computeIfAbsent. This method guarantees atomicity of read and write in case of absence, so two separate threads will get the same instance of Netty HtpMethod class. If containsKey returns true, the following call of get will return the non-null value also without locks.

I can rewrite it like this:

```java
private static io.netty.handler.codec.http.HttpMethod getMethod(String httpMethodName) {
        io.netty.handler.codec.http.HttpMethod value = METHOD_MAP.get(httpMethodName);
        if (value == null) {
            return METHOD_MAP.computeIfAbsent(httpMethodName, io.netty.handler.codec.http.HttpMethod::valueOf);
        }

        return value;
    }
```

That implementation should have no heavy side effects for read and should not introduce any inconsistencies with computeWithAbsent

Reply:
Of course, the simplest variant would be not to use any map like at all. But this could cause performance issues for high load, since netty HttpMethod.valueOf method returns the same instance for standard methods ,but will create a new instance otherwise for each call. That means, if you have a million REPORT requests, you'll get a million of HttpMethod instances. Of course, GC will get rid of them eventually, but I'm worried, it still can cause bottlenecks.